File: Fibonacci.cpp
#include <iostream>
#include <chrono>
#include <vector>
using namespace std;

// Iterative Fibonacci Calculation (with series printing)
void fibonacciIterative(int n) {
    int a = 0, b = 1, fib;
    cout << "Fibonacci series (iterative): ";
    if (n >= 0) cout << a << " ";  // First Fibonacci number
    if (n >= 1) cout << b << " ";  // Second Fibonacci number
    
    for (int i = 2; i <= n; i++) {
        fib = a + b;
        cout << fib << " ";
        a = b;
        b = fib;
    }
    cout << endl;
}

// Recursive Fibonacci Calculation (with series printing)
int fibonacciRecursive(int n) {
    if (n <= 1)
        return n;
    return fibonacciRecursive(n - 1) + fibonacciRecursive(n - 2);
}

void fibonacciRecursiveSeries(int n) {
    cout << "Fibonacci series (recursive): ";
    for (int i = 0; i <= n; i++) {
        cout << fibonacciRecursive(i) << " ";
    }
    cout << endl;
}

// Measure execution time for the given function over multiple runs
double measureExecutionTime(void (*func)(int), int n, int runs) {
    vector<double> durations;
    for (int i = 0; i < runs; i++) {
        auto start = chrono::high_resolution_clock::now();
        func(n);
        auto end = chrono::high_resolution_clock::now();
        chrono::duration<double> duration = end - start;
        durations.push_back(duration.count() * 1000);  // Convert to milliseconds
    }

    // Calculate average duration
    double totalDuration = 0;
    for (double dur : durations) {
        totalDuration += dur;
    }
    return totalDuration / runs;
}

int main() {
    int n, runs = 1000;
    cout << "Enter the position (n) of Fibonacci number to calculate: ";
    cin >> n;

    // Measure execution time for iterative method and print series
    auto start_iterative = chrono::high_resolution_clock::now();
    fibonacciIterative(n);  // Print series for iterative
    auto end_iterative = chrono::high_resolution_clock::now();
    chrono::duration<double> iterative_duration = end_iterative - start_iterative;
    cout << "Execution time (iterative): " << iterative_duration.count() * 1000 << " milliseconds" << endl;

    // Measure execution time for recursive method and print series
    auto start_recursive = chrono::high_resolution_clock::now();
    fibonacciRecursiveSeries(n);  // Print series for recursive
    auto end_recursive = chrono::high_resolution_clock::now();
    chrono::duration<double> recursive_duration = end_recursive - start_recursive;
    cout << "Execution time (recursive): " << recursive_duration.count() * 1000 << " milliseconds" << endl;

    return 0;
}
// Recursive Approach:

// // Time Complexity: O(2*n)
// // Space Complexity: O(n)

// // Iterative Approach:

// // Time Complexity : O(n)
// // Space Complexity : O(1)

File: Fibonacci.py
import time

def fibonacci_recursion(n):
    if n <= 1:
        return n
    return fibonacci_recursion(n - 1) + fibonacci_recursion(n - 2)

def fibonacci_non_recursion(n):
    if n <= 1:
        return n
    a, b = 0, 1
    for _ in range(2, n + 1):
        temp = a + b
        a = b
        b = temp
    return b

def measure_execution_time(func, n):
    start_time = time.perf_counter()  # Use perf_counter for better precision
    result = func(n)                  # Execute the function
    end_time = time.perf_counter()    # End time
    execution_time = end_time - start_time  # Calculate execution time
    return result, execution_time

if __name__ == "__main__":
    n = int(input("Enter the value of n: "))

    # Measure execution time for recursion
    result_rec, time_rec = measure_execution_time(fibonacci_recursion, n)
    print(f"Fibonacci({n}) using recursion:", end=" ")
    for i in range(n):
        print(fibonacci_recursion(i), end=" ")
    print()  # New line after printing the series
    print(f"Execution time (recursion): {time_rec:.10f} seconds")

    # Measure execution time for non-recursion (iterative)
    result_non_rec, time_non_rec = measure_execution_time(fibonacci_non_recursion, n)
    print(f"Fibonacci({n}) using non-recursion:", end=" ")
    for i in range(n):
        print(fibonacci_non_recursion(i), end=" ")
    print()  # New line after printing the series
    print(f"Execution time (non-recursion): {time_non_rec:.10f} seconds")

''' Recursive Approach:

// Time Complexity: O(2*n)
// Space Complexity: O(n)

// Iterative Approach:

// Time Complexity : O(n)
// Space Complexity : O(1)'''